### PR TITLE
feat(sync): track incoming sync requests, allow subscriptions without sync, close inactive replicas

### PR DESCRIPTION
**Pull request by Frando**
## Description

This PR adds a few missing parts for sync in iroh:

* Allow to subscribe to replicas if they're not in the set of syncing replicas (this was not possible before).
* Properly track open replicas in the `LiveSync` and close them once they're not needed anymore. Therefore adds `iroh_sync::store::close_replica` to remove a replica instance from the store-internal cache of open replicas. This is needed because otherwise it grows unbounded (for anchor nodes especially).
* Pass incoming connections to the `LiveSync` actor and handle them there. Before, incoming connections for sync were *always* processed if a replica with the requested `NamespaceId`  is found in the local replica store. This PR changes this mechanism so that incoming sync connections are passed into the sync engine first. This leads to the changes that motivate this PR:
    * Incoming sync requests are only answered if the document (replica) in question was opened for sync before (via a call to `start_sync`). Note that this is *not* persisted at the moment, but should be (separate change though).
    * We can track the sync status for outgoing *and incoming* sync flows. This will reduce the number of sync connections, because currently we will oftenly double-sync in both directions (unneeded, because the sync is always bidirectional). Now, we abort if a sync is already in progress.
* Expose events about successfull and failed sync runs  to the upper layer via the document subscriptions with a new `LiveEvent:.SyncFinished` that reports all available information

## Notes & open questions

Do we want to track history of syncs internally or leave it up to the app via the events? Not sure myself.

The diff of `sync_engine/live.rs` is a little hard to read, because there's a change from `TopicId` to `NamespaceId` in a few places that is mostly unrelated. Sorry for the mixup, but it makes sense to track everything by `NamespaceId` and it ended up in a commit here.

Logging is much improved in this PR. 

Maybe we can improve the code flow a little to make things more obvious.

## Change checklist

- [x] Self-review.
- [x] Documentation updates if relevant.
- [ ] Tests if relevant.

### CODE REVIEW

File: iroh-sync/src/net/codec.rs [hunk 1]
```diff
@@ -260,10 +293,16 @@ mod tests {
...
         let bob_task = tokio::task::spawn(async move {
-            run_bob::<store::memory::Store, _, _>(
+            run_bob::<store::memory::Store, _, _, _, _>(
                 &mut bob_writer,
                 &mut bob_reader,
-                bob_replica_store_task,
+                |namespace, _| {
+                    futures::future::ready(
+                        bob_replica_store_task
+                            .open_replica(&namespace)
```
Comment (dignifiedquire) [inline iroh-sync/src/net/codec.rs:302]: this is a blocking call, we should probably mark it as such

File: iroh-sync/src/net/codec.rs [hunk 2]
```diff
@@ -459,10 +498,12 @@ mod tests {
         let (mut bob_reader, mut bob_writer) = tokio::io::split(bob);
         let bob_store = bob_store.clone();
         let bob_task = tokio::task::spawn(async move {
-            run_bob::<S, _, _>(
+            run_bob::<S, _, _, _, _>(
                 &mut bob_writer,
                 &mut bob_reader,
-                bob_store,
+                |namespace, _| {
+                    futures::future::ready(bob_store.open_replica(&namespace).map(Into::into))
```
Comment (dignifiedquire) [inline iroh-sync/src/net/codec.rs:505]: same here

File: iroh/src/commands/sync.rs [hunk 3]
```diff
@@ -247,6 +247,9 @@ impl DocCommands {
                         LiveEvent::ContentReady { hash } => {
                             println!("content ready: {}", fmt_short(hash.as_bytes()))
                         }
+                        LiveEvent::SyncFinished(details) => {
+                            println!("sync finished: {details:?}")
```
Comment (dignifiedquire) [inline iroh/src/commands/sync.rs:251]: maybe `sync finished for {fmt_short(..)}`?

**Review by dignifiedquire — APPROVED** [review]

File: iroh-sync/src/net/codec.rs (hunk 1 above)
Comment (Frando) [reply iroh-sync/src/net/codec.rs:302]: deferred for later: https://github.com/n0-computer/iroh/issues/1502

File: iroh-sync/src/net/codec.rs (hunk 2 above)
Comment (Frando) [reply iroh-sync/src/net/codec.rs:505]: deferred to https://github.com/n0-computer/iroh/issues/1502

**Pull request merged**